File: generated_codes/in-context/GPT-5.5/cpp/10.cpp
#pragma once
#include "../utils/SetADT.hpp"
#include <atomic>
#include <climits>

/**
 * Lock-free Binary Search Tree with CAS-based insertion
 * and CAS-mark logical deletion.
 *
 * Design: no physical unlinking — marked nodes remain in structure.
 * This keeps deletion simple and avoids unsafe memory reclamation issues.
 *
 * Linearization:
 *   add():      CAS on parent child pointer OR CAS marked true->false
 *   remove():   CAS on node->marked false->true
 *   contains(): read of node->marked
 */
class LockFreeBST : public SetADT {
private:
    struct Node {
        int key;
        std::atomic<bool> marked;
        std::atomic<Node*> left;
        std::atomic<Node*> right;

        Node(int k)
            : key(k), marked(false), left(nullptr), right(nullptr) {}
    };

    Node* root;

    void deleteSubtree(Node* node) {
        if (!node) return;
        Node* l = node->left.load(std::memory_order_relaxed);
        Node* r = node->right.load(std::memory_order_relaxed);
        deleteSubtree(l);
        deleteSubtree(r);
        delete node;
    }

    Node* findNode(int key) const {
        Node* curr = root->right.load(std::memory_order_acquire);

        while (curr) {
            if (key == curr->key) {
                return curr;
            } else if (key < curr->key) {
                curr = curr->left.load(std::memory_order_acquire);
            } else {
                curr = curr->right.load(std::memory_order_acquire);
            }
        }

        return nullptr;
    }

public:
    LockFreeBST() {
        root = new Node(INT_MIN);
    }

    ~LockFreeBST() override {
        deleteSubtree(root);
    }

    bool contains(int key) override {
        Node* node = findNode(key);
        return node && !node->marked.load(std::memory_order_acquire);
    }

    bool add(int key) override {
        while (true) {
            Node* parent = root;
            Node* curr = root->right.load(std::memory_order_acquire);
            bool goLeft = false;

            while (curr) {
                if (key == curr->key) {
                    bool expected = true;
                    if (curr->marked.compare_exchange_strong(
                            expected, false,
                            std::memory_order_acq_rel,
                            std::memory_order_acquire)) {
                        return true;
                    }

                    if (!curr->marked.load(std::memory_order_acquire)) {
                        return false;
                    }

                    continue;
                }

                parent = curr;

                if (key < curr->key) {
                    goLeft = true;
                    curr = curr->left.load(std::memory_order_acquire);
                } else {
                    goLeft = false;
                    curr = curr->right.load(std::memory_order_acquire);
                }
            }

            Node* newNode = new Node(key);
            std::atomic<Node*>& child =
                (parent == root || !goLeft) ? parent->right : parent->left;

            Node* expected = nullptr;
            if (child.compare_exchange_strong(
                    expected, newNode,
                    std::memory_order_acq_rel,
                    std::memory_order_acquire)) {
                return true;
            }

            delete newNode;
        }
    }

    bool remove(int key) override {
        Node* node = findNode(key);
        if (!node) return false;

        bool expected = false;
        if (!node->marked.compare_exchange_strong(
                expected, true,
                std::memory_order_acq_rel,
                std::memory_order_acquire)) {
            return false;
        }

        // Node has been marked
        return true;
    }
};

File: generated_codes/in-context/GPT-5.5/java/10.java
package com.example.Sets;

import com.example.utils.SetADT;
import java.util.concurrent.atomic.AtomicReference;

/**
 * Lock-free Binary Search Tree using immutable path-copying.
 *
 * Linearization points:
 * - add(): successful CAS on root
 * - remove(): successful CAS on root
 * - contains(): atomic root snapshot read
 *
 * No locks are used. Failed CAS means another thread completed an update.
 */
public class ConcurrentDataStructure implements SetADT {

    private static final class Node {
        final int key;
        final Node left;
        final Node right;

        Node(int key, Node left, Node right) {
            this.key = key;
            this.left = left;
            this.right = right;
        }
    }

    private final AtomicReference<Node> root;

    public ConcurrentDataStructure() {
        root = new AtomicReference<>(null);
    }

    @Override
    public boolean add(int key) {
        while (true) {
            Node oldRoot = root.get();

            if (contains(oldRoot, key)) {
                return false;
            }

            Node newRoot = insert(oldRoot, key);

            if (root.compareAndSet(oldRoot, newRoot)) {
                return true;
            }
        }
    }

    @Override
    public boolean remove(int key) {
        while (true) {
            Node oldRoot = root.get();

            if (!contains(oldRoot, key)) {
                return false;
            }

            Node newRoot = delete(oldRoot, key);

            if (root.compareAndSet(oldRoot, newRoot)) {
                // Node has been marked
                return true;
            }
        }
    }

    @Override
    public boolean contains(int key) {
        return contains(root.get(), key);
    }

    private boolean contains(Node node, int key) {
        Node curr = node;

        while (curr != null) {
            if (key < curr.key) {
                curr = curr.left;
            } else if (key > curr.key) {
                curr = curr.right;
            } else {
                return true;
            }
        }

        return false;
    }

    private Node insert(Node node, int key) {
        if (node == null) {
            return new Node(key, null, null);
        }

        if (key < node.key) {
            return new Node(node.key, insert(node.left, key), node.right);
        }

        if (key > node.key) {
            return new Node(node.key, node.left, insert(node.right, key));
        }

        return node;
    }

    private Node delete(Node node, int key) {
        if (node == null) {
            return null;
        }

        if (key < node.key) {
            return new Node(node.key, delete(node.left, key), node.right);
        }

        if (key > node.key) {
            return new Node(node.key, node.left, delete(node.right, key));
        }

        if (node.left == null) {
            return node.right;
        }

        if (node.right == null) {
            return node.left;
        }

        Node successor = min(node.right);
        Node newRight = deleteMin(node.right);

        return new Node(successor.key, node.left, newRight);
    }

    private Node min(Node node) {
        Node curr = node;

        while (curr.left != null) {
            curr = curr.left;
        }

        return curr;
    }

    private Node deleteMin(Node node) {
        if (node.left == null) {
            return node.right;
        }

        return new Node(node.key, deleteMin(node.left), node.right);
    }
}